executor/insert_common.go

```diff
@@ -455,33 +457,126 @@ func (e *InsertValues) setValueForRefColumn(row []types.Datum, hasValue []bool)
 	// process `insert|replace into ... select ... from ...`
 	e := base.insertCommon()
 	selectExec := e.children[0]
 	fields := retTypes(selectExec)
 	chk := tryNewCacheChunk(selectExec)
 	iter := chunk.NewIterator4Chunk(chk)
 	rows := make([][]types.Datum, 0, chk.Capacity())
+	var chunkPool *sync.Pool
 
 	sessVars := e.ctx.GetSessionVars()
 	batchSize := sessVars.DMLBatchSize
 	batchInsert := sessVars.BatchInsert && !sessVars.InTxn() && variable.EnableBatchDML.Load() && batchSize > 0
 	memUsageOfRows := int64(0)
 	memUsageOfExtraCols := int64(0)
 	memTracker := e.memTracker
+
+	var (
+		dataCh chan *chunk.Chunk
+	)
+	etlConcurrency, etlBatchSize := sessVars.ETLConcurrency, sessVars.ETLBatchSize
+	if etlConcurrency <= 1 {
+		etlMode = false
+	}
+	if etlMode {
+		var workers []insertCommon
+		workers, err = createFromSelectWorkers(ctx, base, etlConcurrency)
+		if err != nil {
+			return err
+		} else if len(workers) == 0 {
+			// executor not support, fallback to single thread.
+			etlMode = false
+		} else {
+			affectedRow := uint64(0)
+			maxCap := e.maxChunkSize
+			approximateChannelSize := 2 * etlConcurrency * (batchSize/maxCap + 1)
+			dataCh = make(chan *chunk.Chunk, approximateChannelSize)
+			chunkBackCh := make(chan *chunk.Chunk, approximateChannelSize)
+			var eg *errgroup.Group
+			eg, ctx = errgroup.WithContext(ctx)
+
+			chunkPool = &sync.Pool{
+				New: func() any {
+					return chunk.New(fields, maxCap, maxCap)
+				},
+			}
+			go func() {
+				for {
+					select {
+					case c := <-chunkBackCh:
+						if c != nil {
+							chunkPool.Put(c)
+						} else {
+							return
+						}
+					case <-ctx.Done():
+						return
+					}
+				}
+			}()
+
+			defer func() {
+				err = eg.Wait()
+				for range dataCh {
+				}
+				close(chunkBackCh)
+				for range chunkBackCh {
+				}
+				sessVars.StmtCtx.SetAffectedRows(affectedRow)
+				for _, worker := range workers {
+					switch w := worker.(type) {
+					case *InsertExec:
+						e.stats.Merge(w.stats)
+						e.lastInsertID = w.lastInsertID
+						e.runtimeStats.Merge(w.runtimeStats)
+					default:
+					}
+				}
+			}()
+			logutil.Logger(ctx).Info("run parallel insert",
+				zap.Int("concurrency", etlConcurrency),
+				zap.Int("batch size", etlBatchSize))
+			for i := 0; i < etlConcurrency; i++ {
+				worker := workers[i]
+				eg.Go(func() error {
+					return insertRowsFromSelectWorker(ctx, worker, batchSize, dataCh, chunkBackCh, &affectedRow, &sessVars.Killed)
+				})
+			}
+		}
+	}
+
 	extraColsInSel := make([][]types.Datum, 0, chk.Capacity())
 	// In order to ensure the correctness of the `transaction write throughput` SLI statistics,
 	// just ignore the transaction which contain `insert|replace into ... select ... from ...` statement.
 	e.ctx.GetTxnWriteThroughputSLI().SetInvalid()
+LOOP:
 	for {
+		// we send the datum to other goroutine, and the datum has reference to the chunk,
+		// for safety, create a new chunk.
+		if etlMode {
+			chk = chunkPool.Get().(*chunk.Chunk)
```

Comment:
How about using `chunkBackCh` directly? We can send enough chunks to `chunkBackCh` first.

Reply:
It's hard to decide how much is enough...